File: app/src/main/java/tr/name/fatihdogan/books/view/AuthorView.java
package tr.name.fatihdogan.books.view;

import android.content.Context;
import android.content.Intent;
import android.content.res.TypedArray;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.CardView;
import android.support.v7.widget.PopupMenu;
import android.util.AttributeSet;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import tr.name.fatihdogan.books.R;
import tr.name.fatihdogan.books.activity.MainActivity;
import tr.name.fatihdogan.books.repository.AppDatabase;
import tr.name.fatihdogan.books.utils.EditTextUtils;
import tr.name.fatihdogan.books.utils.ThreadUtils;

public class AuthorView extends CardView implements View.OnClickListener, PopupMenu.OnMenuItemClickListener {

    private static final String TOOLS_NS = "http://schemas.android.com/tools";

    private TextView nameTextView;
    private ImageButton optionButton;
    private PopupMenu optionPopupMenu;
    private MenuItem editMenuItem;

    public AuthorView(Context context) {
        super(context);
        init(context, null);

    }

    public AuthorView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);

    }

    public AuthorView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, @Nullable AttributeSet attrs) {
        setRadius(5);
        ViewCompat.setElevation(this, 5);
        inflate(context, R.layout.view_authorview, this);
        setUseCompatPadding(true);
        nameTextView = findViewById(R.id.name_text_view);
        optionButton = findViewById(R.id.option_button);
        setOnClickListener(this);
        optionButton.setOnClickListener(this);
        optionPopupMenu = new PopupMenu(context, optionButton);
        editMenuItem = optionPopupMenu.getMenu().add(R.string.edit);
        optionPopupMenu.setOnMenuItemClickListener(this);
        if (attrs != null) {
            TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.AuthorView);
            setName(a.getText(R.styleable.AuthorView_name));
            a.recycle();
        }

        if (isInEditMode() && attrs != null) {
            setName(attrs.getAttributeValue(TOOLS_NS, "name"));
        }
    }

    public void setName(CharSequence name) {
        nameTextView.setText(name);
    }

    public CharSequence getName() {
        return nameTextView.getText();
    }

    @Override
    public void onClick(View v) {
        if (v == this) {
            Intent intent = new Intent(v.getContext(), MainActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
            intent.putExtra("AUTHOR", nameTextView.getText());
            v.getContext().startActivity(intent);
        }
        if (v == optionButton) {
            optionPopupMenu.show();
        }
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        if (item == editMenuItem) {
            showEditDialog();
            return true;
        }
        return false;
    }

    private void showEditDialog() {
        final AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setTitle(R.string.edit_author_name);

        final EditText input = new EditText(getContext());
        input.setSingleLine();
        LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT);
        input.setText(getName());
        input.setLayoutParams(lp);
        builder.setView(input);
        builder.setPositiveButton(android.R.string.ok, (dialog, which) ->
                ThreadUtils.runOnBackground(() ->
                        AppDatabase.getBookDao().renameAuthor(
                                nameTextView.getText().toString(), input.getText().toString())));
        builder.setNegativeButton(android.R.string.cancel, (dialog, which) -> dialog.cancel());
        builder.show();

        EditTextUtils.focusAndShowKeyboard(input);
    }

}

File: app/src/main/java/com/fthdgn/books/view/AuthorView.kt
package com.fthdgn.books.view

import android.content.Context
import android.content.Intent
import android.support.v4.view.ViewCompat
import android.support.v7.app.AlertDialog
import android.support.v7.widget.CardView
import android.support.v7.widget.PopupMenu
import android.util.AttributeSet
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import com.fthdgn.books.BaseApplication
import com.fthdgn.books.R
import com.fthdgn.books.activity.MainActivity
import com.fthdgn.books.repository.BookDao
import com.fthdgn.books.utils.bind
import com.fthdgn.books.utils.focusAndShowKeyboard
import com.fthdgn.books.utils.runBackground

class AuthorView : CardView, View.OnClickListener, PopupMenu.OnMenuItemClickListener {

    private val bookDao: BookDao = BaseApplication.getAppComponent().bookDao()

    private val nameTextView: TextView by bind(R.id.name_text_view)
    private val optionButton: ImageButton by bind(R.id.option_button)
    private val optionPopupMenu: PopupMenu by lazy { PopupMenu(context, optionButton) }
    private lateinit var editMenuItem: MenuItem

    internal var name: CharSequence
        get() = nameTextView.text
        set(name) {
            nameTextView.text = name
        }

    constructor(context: Context) : super(context) {
        init(context, null)

    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context, attrs)

    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        radius = 5f
        ViewCompat.setElevation(this, 5f)
        View.inflate(context, R.layout.view_authorview, this)
        useCompatPadding = true
        setOnClickListener(this)
        optionButton.setOnClickListener(this)
        optionPopupMenu.setOnMenuItemClickListener(this)
        editMenuItem = optionPopupMenu.menu.add(R.string.edit)

        if (attrs != null) {
            val a = context.obtainStyledAttributes(attrs, R.styleable.AuthorView)
            name = a.getText(R.styleable.AuthorView_name)
            a.recycle()
        }

        if (isInEditMode && attrs != null) {
            name = attrs.getAttributeValue(TOOLS_NS, "name")
        }
    }

    override fun onClick(v: View) {
        if (v === this) {
            val intent = Intent(v.getContext(), MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
            intent.putExtra("AUTHOR", nameTextView.text)
            v.getContext().startActivity(intent)
        }
        if (v === optionButton) {
            optionPopupMenu.show()
        }
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        if (item === editMenuItem) {
            showEditDialog()
            return true
        }
        return false
    }

    private fun showEditDialog() {
        val builder = AlertDialog.Builder(context)
        builder.setTitle(R.string.edit_author_name)

        val input = EditText(context)
        input.setSingleLine()
        val lp = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT)
        input.setText(name)
        input.layoutParams = lp
        builder.setView(input)
        builder.setPositiveButton(android.R.string.ok) { _, _ ->
            runBackground {
                bookDao.renameAuthor(nameTextView.text.toString(), input.text.toString())
            }
        }
        builder.setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.cancel() }
        builder.show()
        input.focusAndShowKeyboard()
    }

    companion object {

        private val TOOLS_NS = "http://schemas.android.com/tools"
    }

}
